Seminar 8: DNA Methylation Analysis (Array)
========================================================
by Erica Acton

```{r include=FALSE}
library(GEOquery)
library(AnnotationForge)
library(wateRmelon)
library(IlluminaHumanMethylation450k.db)
library(limma)
library(gplots)
```
Explore 450K methylation
------------------------------------------------------------------------------
>Retrieve datasets.
```{r}
if(file.exists("methyl_ALL.Rdata")) {
  load("methyl_ALL.Rdata")
} else {
  GSE39141 <- getGEO("GSE39141") #29 ALL and 4 healthy B cells
  show(GSE39141)
  GSE42865 <- getGEO("GSE42865") #9 heathy and 7 other conditions
  show(GSE42865)
  #Extract expression matrices and convert to data frames
  ALL.dat <- as.data.frame(exprs(GSE39141[[1]]))
  CTRL.dat <- as.data.frame(exprs(GSE42865[[1]]))
  #Obtain meta-data
  ALL.meta <- pData(phenoData(GSE39141[[1]]))
  CTRL.meta <- pData(phenoData(GSE42865[[1]]))
  #Create labels
  ALL.meta$Group <- c(rep("ALL",29), rep("HBC",4))
  #subset meta-data and data for healthy donors
   CTRL.meta <- droplevels(subset(CTRL.meta,grepl("Healthy donor", characteristics_ch1.1)))
  CTRL.dat <- subset(CTRL.dat, select=as.character(CTRL.meta$geo_accession))
  #Rename variables
  names(ALL.dat) <- paste(ALL.meta$Group, gsub("GSM","", names(ALL.dat)), sep="_")
  names(CTRL.dat) <- paste("HBC", gsub("GSM","", names(CTRL.dat)), sep="_")
  
  save(ALL.dat, CTRL.dat, ALL.meta, CTRL.meta, file="methyl_ALL.Rdata")
}
```
>Exploratory analysis looking at the density of Beta values for probes in the two datasets.
```{r}
ALL.mean <- rowMeans(ALL.dat, na.rm = T)
CTRL.mean <- rowMeans(CTRL.dat, na.rm = T)
both.mean <- data.frame(Beta = c(ALL.mean, CTRL.mean), type = c(rep("ALL", 
    length(ALL.mean)), rep("Control", length(CTRL.mean))))

densityplot(~Beta, both.mean, groups = type, grid = TRUE, plot.points = FALSE, 
    auto.key = TRUE, main="Density of Beta values before normalization")
```
Normalization
----------------------------------------------------------------------------------
>Combine data from both experiments into one matrix.
```{r}
beta.matrix <- as.matrix(cbind(ALL.dat,CTRL.dat))
str(beta.matrix, max.level=0)
```
>Perform quantile normalization of beta values.
```{r}
system.time(beta.norm <- betaqn(beta.matrix))
#takes longer than Gloria...poor computer...
```
>Plot density of Beta values after normalization.
```{r}
beta.qn <- data.frame(Beta=c(rowMeans(beta.norm[,colnames(ALL.dat)], na.rm=T),
                rowMeans(beta.norm[,colnames(CTRL.dat)], na.rm=T)),
                group=c(rep("ALL",nrow(ALL.dat)), rep("Control",nrow(CTRL.dat))))

densityplot(~Beta, beta.qn, groups = group, grid = TRUE, plot.points = FALSE, 
    auto.key = TRUE, main="Density of Beta values after normalization")
```
M Values
-------------------------------------------------------------------------------
>Apply transformation ie. compute M value.
```{r}
M.norm <- beta2m(beta.norm)
```
CpG Islands
-------------------------------------------------------------------------
>Extract probe ID to CpG islands association.
```{r}
cginame <- as.data.frame(IlluminaHumanMethylation450kCPGINAME)
names(cginame) <- c("Probe_ID", "cginame")
rownames(cginame) <- cginame$Probe_ID
length(levels(factor(cginame$cginame)))
```
>Restrict probes to within CGIs.
```{r}
beta.inCGI <- beta.norm[cginame$Probe_ID, ]
M.inCGI <- M.norm[cginame$Probe_ID, ]
nrow(M.inCGI)
```
>Aggregate probes to CGIs (B-values).
```{r}
beta.CGI <- aggregate(beta.inCGI, by = list(cginame$cginame), mean, na.rm = T)
rownames(beta.CGI) <- beta.CGI[, "Group.1"]
beta.CGI <- subset(beta.CGI, select = -Group.1)
str(beta.CGI, max.level = 0)
```
>Aggregate probes to CGIs (M-values).
```{r}
M.CGI <- aggregate(M.inCGI, by = list(cginame$cginame), mean, na.rm = T)
rownames(M.CGI) <- M.CGI[, "Group.1"]
M.CGI <- subset(M.CGI, select = -Group.1)
str(M.CGI, max.level = 0)
```
>Boxplot of distribution of CGI M-values.
```{r}
boxplot(M.CGI, xlab="Samples", ylab="M values", main="Distribution of CGI M values")
```
Differential methylation analysis with limma
---------------------------------------------------------------------------------
>Use a linear model to identify differentially methylated CGIs. Design matrix, lmFit, eBayes, topTable...
```{r}
design <- data.frame(Group = relevel(factor(gsub("_[0-9]+", "", colnames(M.CGI))), ref = "HBC"), row.names = colnames(M.CGI))
str(design)

(DesMat <- model.matrix(~Group, design))

DMRfit <- lmFit(M.CGI, DesMat)
DMRfitEb <- eBayes(DMRfit)
cutoff <- 0.01
DMR <- topTable(DMRfitEb, coef = "GroupALL", number = Inf, p.value = cutoff)
head(DMR)  
#Number of CGIs differentially methylated between ALL and Control group at a cutoff of FDR=0.01
nrow(DMR)
```
>Make a heatmap of the top 100 hits.
```{r}
DMR100 <- topTable(DMRfitEb, coef = "GroupALL", number = 100)
DMR.CGI <- t(as.matrix(subset(beta.CGI, rownames(beta.CGI) %in% rownames(DMR100))))
str(DMR.CGI, max.level = 0)

col <- c(rep("darkgoldenrod1", times = nrow(DMR.CGI)))
col[grepl("HBC", rownames(DMR.CGI))] <- "forestgreen"
op <- par(mai = rep(0.5, 4))
heatmap.2(DMR.CGI, col = redblue(256), RowSideColors = col, density.info = "none", trace = "none", Rowv = TRUE, Colv = TRUE, labCol = FALSE, labRow = FALSE, 
dendrogram = "row", margins = c(1, 5))
legend("topright", c("ALL", "HBC"), col = c("darkgoldenrod1", "forestgreen"), 
    pch = 15)
par(op)
```
>Make a stripplot of beta values of probes within top 5 CGI hits.
```{r}
DMR5 <- topTable(DMRfitEb, coef = "GroupALL", number = 5)
beta.DMR5probe <- beta.inCGI[cginame[rownames(beta.inCGI), ]$cginame %in% rownames(DMR5), 
    ]
beta.DMR5probe.tall <- melt(beta.DMR5probe, value.name = "M", varnames = c("Probe_ID", 
    "Sample"))
beta.DMR5probe.tall$Group <- factor(gsub("_[0-9]+", "", beta.DMR5probe.tall$Sample))
beta.DMR5probe.tall$CGI <- factor(cginame[as.character(beta.DMR5probe.tall$Probe_ID), 
    ]$cginame)
(beta.DMR5.stripplot <- ggplot(data = beta.DMR5probe.tall, aes(x = Group, y = M, 
    color = Group)) + geom_point(position = position_jitter(width = 0.05), na.rm = T) + stat_summary(fun.y = mean, aes(group = 1), geom = "line", color = "black") + facet_grid(. ~ CGI) + ggtitle("Probe beta values within top 5 DM CGIs") + xlab("Group") + ylab("Beta") + theme_bw())
```
>Plot the location of DM probes along each chromosome.
```{r}
#chromosome lengths 1-22 and X
chrlen <- unlist(as.list(IlluminaHumanMethylation450kCHRLENGTHS)[c(as.character(1:22), "X")])
chrlen <- data.frame(chr = factor(names(chrlen)), length = chrlen)
chr <- IlluminaHumanMethylation450kCHR  
#probe identifiers mapped to chromosome
chr <- unlist(as.list(chr[mappedkeys(chr)]))
#coordinate of probe
coord <- IlluminaHumanMethylation450kCPGCOORDINATE
#probe identifiers mapped to coordinate
coord <- unlist(as.list(coord[mappedkeys(coord)]))
coord <- data.frame(chr = chr[intersect(names(chr), names(coord))], coord = coord[intersect(names(chr), names(coord))])
# coordinates of probes in DM CGIs
coordDMRprobe <- droplevels(na.omit(coord[cginame[cginame$cginame %in% rownames(DMR),]$Probe_ID, ]))
#plot
(coord.plot <- ggplot(data = coordDMRprobe) + geom_linerange(aes(factor(chr, 
  levels = c("X", as.character(22:1))), ymin = 0, ymax = length), data = chrlen, 
    alpha = 0.5) + geom_point(aes(x = factor(chr, levels = c("X", as.character(22:1))), y = coord), position = position_jitter(width = 0.03), na.rm = T) + ggtitle("DMR positions on chromosomes") + ylab("Position of DMRs") + xlab("chr") + coord_flip() + theme_bw())
```
Interpretation and functional enrichment analysis
-----------------------------------------------------------------------
>Associate DMRs with biological features/functions by associating them with genes.  Get annotation information ie. probes associated with different genomic regions or gene IDs.  Below is a list of objects available in the package to acheive this task.

```{r}
ls("package:IlluminaHumanMethylation450k.db")
```
