Seminar 3: Introduction to R Graphics (Lattice + ggplot2)
===========================================================
by Erica Acton

Lattice
---------------------------------------------------------------------------------

>Load data.
```{r}
library(lattice)
library(grid)
library(hexbin)

kDat <- readRDS("/home/eacton/R/stat540-2014-ACTON-ERICA/GSE4051_MINI.rds")
str(kDat)
table(kDat$devStage)
table(kDat$gType)
with(kDat, table(devStage,gType))
```

Scatterplots.
------------------------------------------------------------------------------
>Plot gene expression of one gene against another.
```{r}
xyplot(eggBomb~crabHammer, kDat)
# You try!
xyplot(poisonFang~crabHammer, kDat)
```
>Plot the gene expression of one gene against two others at the same time.
```{r}
xyplot(eggBomb + poisonFang~crabHammer, kDat,auto.key=TRUE)
```
>Separate the two genes instead via a grid.
```{r}
xyplot(eggBomb + poisonFang~crabHammer, kDat, outer=TRUE, grid=TRUE)
```
>Add genotype information.
```{r}
xyplot(eggBomb + poisonFang ~ crabHammer, kDat, outer=TRUE, grid=TRUE, groups=gType, auto.key=TRUE)
```
>Reshape data.
```{r}
nDat <- with(kDat, data.frame(sidChar,sidNum,devStage,gType,crabHammer, probeset=factor(rep(c("eggBomb", "poisonFang"), each=nrow(kDat))), geneExp=c(eggBomb,poisonFang)))
str(nDat)
```
>Remake previous plots with panel syntax.
```{r}
xyplot(geneExp~crabHammer|probeset, nDat, grid=TRUE, groups=gType, auto.key=TRUE)
#You try!
xyplot(geneExp~crabHammer|probeset, nDat, grid=TRUE, groups=devStage, auto.key=TRUE)
```
Stripplots
----------------------------------------------------------------------------------------------------------------------------------------------
>Futher data reshaping.
```{r}
oDat <- with(kDat, data.frame(sidChar,sidNum,devStage,gType,probeset=factor(rep(c("crabHammer", "eggBomb", "poisonFang"), each=nrow(kDat))), geneExp=c(crabHammer,eggBomb,poisonFang)))
str(oDat)
```
>Univariate stripplot of 1 gene.
```{r}
stripplot(~geneExp,oDat)
```
>Stripplot of 3 genes, and then 3 genes with jitter spacing.
```{r}
stripplot(probeset~geneExp,oDat)
stripplot(probeset~geneExp,oDat, jitter.data=TRUE)
```
>Stripplot of genes separated by panels.
```{r}
stripplot(~geneExp|probeset, oDat, layout=c(nlevels(oDat$probeset),1))
```
>Stripplot of genes separated by panels with genotype information.
```{r}
stripplot(~geneExp|probeset, oDat, layout=c(nlevels(oDat$probeset),1), groups=gType, auto.key=TRUE)
```
>Gene expression over the course of development.
```{r}
stripplot(geneExp~devStage,oDat)
```
>Gene expression over the course of development for each of the 3 genes.
```{r}
stripplot(geneExp~devStage|probeset,oDat, layout=c(nlevels(oDat$probeset),1))
```
>Gene expression over the course of development for each of the 3 genes with genotype information.
```{r}
stripplot(geneExp~devStage|probeset,oDat, layout=c(nlevels(oDat$probeset),1),groups=gType, auto.key=TRUE)
```
>Gene expression over the course of development for each of the 3 genes with genotype information and the addition of averages.
```{r}
stripplot(geneExp~devStage|probeset,oDat, layout=c(nlevels(oDat$probeset),1),groups=gType, auto.key=TRUE, grid=TRUE, type=c('p','a'))
```
Densityplots
-----------------------------------------------------------------------------------------------------------------------------------------
>Density plot of gene expression.
```{r}
densityplot(~geneExp, oDat)
```
>Density plot of gene expression with genotype displayed in separate panels.
```{r}
densityplot(~geneExp|gType, oDat, grid=TRUE)
```
>Density plot of gene expression with genotype displayed by colour.
```{r}
densityplot(~geneExp, oDat, groups=gType, auto.key=TRUE)
```
>Density plot of gene expression with genotype displayed by colour displaying bandwidth.
```{r}
Bw <- 0.2
n <- 400
densityplot(~geneExp, oDat,groups=gType, auto.key=TRUE, bw=Bw, n=n,
            main=paste("bw =",Bw, ", n =",n))
#You try!
Bw2 <- 0.4
n2 <- 200
densityplot(~geneExp, oDat,groups=devStage, auto.key=TRUE, bw=Bw2, n=n2,
            main=paste("bw =",Bw, ", n =",n))
```
Boxplots
-------------------------------------------------------------------------------------------------------------------------------------------
>Boxplot of gene expression and developmental stage.
```{r}
bwplot(geneExp~devStage, oDat)
```
>Boxplot of gene expression and developmental stage by genotype
```{r}
bwplot(geneExp~devStage|gType, oDat)
```
>Violin plot.
```{r}
bwplot(geneExp~devStage|gType, oDat, panel=panel.violin)
```
Heatmaps
----------------------------------------------------------------------------------------------------------------------------------------------
>Load larger dataset.
```{r}
prDat <- read.table("/home/eacton/R/stat540-2014-ACTON-ERICA/GSE4051_data.tsv")
str(prDat, max.level=0)
prDes <- readRDS("/home/eacton/R/stat540-2014-ACTON-ERICA/GSE4051_design.rds")
str(prDes)
```
>Choose a random sample size of 50 probes in a repeatable way. Convert to matrix and add meaningful labels.

```{r}
set.seed(1)
(yo <- sample(1:nrow(prDat), size=50))
hDat <- prDat[yo,]
str(hDat)

hDat <- as.matrix(t(hDat))
rownames(hDat) <- with(prDes, paste(devStage, gType, sidChar, sep="_"))
str(hDat)
```
>Make a heatmap. Immediately make a heatmap with better colours.
```{r}
heatmap(hDat, Rowv=NA, Colv=NA,scale="none",margins=c(5,8))
heatmap(hDat, Rowv=NA, Colv=NA, col = cm.colors(256),scale="none",margins=c(5,8))
```
>Load RColorBrewer and try the grey and blue/purple palettes.
```{r}
library(RColorBrewer)
display.brewer.all()
jGraysFun <- colorRampPalette(brewer.pal(n=9, "Greys"))
jBuPuFun <- colorRampPalette(brewer.pal(n=9, "BuPu"))
heatmap(hDat, Rowv=NA, Colv=NA, col = jGraysFun(256),scale="none",margins=c(5,8))
heatmap(hDat, Rowv=NA, Colv=NA, col = jBuPuFun(256),scale="none",margins=c(5,8))
```
>Add dendrograms.  
```{r}
heatmap(hDat, col = jBuPuFun(256),margins=c(5,8))
```
>Scale according to column.
```{r}
heatmap(hDat, col = jBuPuFun(256),margins=c(5,8), scale=c("column"))
```
Overplotting
-------------------------------------------------------------------------------------------------------------------------------------------
>Select 2 random, but repeatable, samples to plot against each other.
```{r}
set.seed(924)
(yo <- sample(1:ncol(prDat), size=2))
y <- prDat[[yo[1]]]
z <- prDat[[yo[2]]]
str(y)
str(z)
xyplot(y~z, asp=1)
```
>Explore the smoothscatter plotting function, or by specifying in 'panel' from lattice.
```{r}
smoothScatter(y~z, asp=1)
xyplot(y~z, asp=1, panel=panel.smoothScatter, nbin=150)
```
>Explore hexagonal binning.
```{r}
hexbinplot(y~z)
```
Plot Matrix
---------------------------------------------------------------------------------------------------------------------------------------
>Take a larger sample of columns. Create pairwise plots.
```{r}
set.seed(3)
(yo <- sample(1:ncol(prDat), size=4))
pairDat <- subset(prDat,select=yo)
str(pairDat)
pairs(pairDat)
```
>Add smoothscatter.
```{r}
pairs(pairDat, panel=function(...) smoothScatter(...,add=TRUE))
```
>Explore splom() and splom() with smoothscatter. This is useful for high-volume scatterplotting.
```{r}
splom(pairDat)
splom(pairDat, panel=panel.smoothScatter, raster=TRUE)
```
>Explore hexplom().
```{r}
hexplom(pairDat)
```
ggplot2
================================================================================
>Load and explore basic terminology.
```{r}
library(ggplot2)
apropos("^geom_")
apropos("^stat_")
apropos("^scale_")
```
>Load data.  Add a few tables for info about data set.
```{r}
kDat <- readRDS("/home/eacton/R/stat540-2014-ACTON-ERICA/GSE4051_MINI.rds")
str(kDat)

table(kDat$devStage)
table(kDat$gType)
with(kDat, table(devStage, gType))
```
qplot
-------------------------------------------------------------------------------
>Make a (basic) qplot.
```{r}
qplot(crabHammer,eggBomb, data=kDat)
```
>Make a scatterplot.  Information is added to ggplot2 in layers.  Such as labels,
titles, and smoothing lines.
```{r}
p <- ggplot(kDat, aes(x=crabHammer, y=eggBomb))
str(p)

(p <- p + geom_point())
(p <- p+ stat_smooth())
(p <- p + theme_bw() + xlab("Expression of crabHammer")+ ylab("Expression of eggBomb")+ ggtitle("Scatterplot for expression levels"))
```
>Reshape data.
```{r}
nDat <-with(kDat, data.frame(sidChar, sidNum, devStage, gType, crabHammer,
      probeset = factor(rep(c("eggBomb", "poisonFang"), each = nrow(kDat))), geneExp = c(eggBomb, poisonFang)))
str(nDat)
```
>Gene expression of poisonFang and eggBomb against crabHammer using color.
```{r}
(p <- ggplot(nDat, aes(crabHammer, geneExp, color = probeset)) + 
   geom_point())
```
>Add a smoothing line.
```{r}
(p <- ggplot(nDat, aes(crabHammer, geneExp, color = probeset)) + 
   geom_point() + stat_smooth(se = F))
```
>Change the line so that it no longer defining a line for each y-axis probeset,
but for the entire set.
```{r}
(p <- ggplot(nDat, aes(crabHammer, geneExp, color = probeset)) + 
   geom_point() + stat_smooth(se = F, aes(group = 1)))
```
>Use facetting to plot probesets in separate panels.
```{r}
(p <- ggplot(nDat, aes(crabHammer, geneExp)) + geom_point() + 
   facet_wrap(~probeset))
```
>Now color can be used to display genotype.
```{r}
(p <- ggplot(nDat, aes(crabHammer, geneExp, color = gType)) + 
   geom_point() + facet_wrap(~ probeset))

#You try! Use color for developmental stage instead of genotype!
(p <- ggplot(nDat, aes(crabHammer, geneExp, color = devStage)) + 
   geom_point() + facet_wrap(~ probeset))
```
Stripplots
--------------------------------------------------------------------------------
>Reshape data.
```{r}
oDat <-with(kDat, data.frame(sidChar, sidNum, devStage, gType,
      probeset = factor(rep(c("crabHammer", "eggBomb","poisonFang"), each = nrow  (kDat))), geneExp = c(crabHammer, eggBomb, poisonFang)))
str(oDat)
```
>Plot the expression level of each gene. Then add jitter for spacing.
```{r}
(p <- ggplot(oDat, aes(geneExp, probeset)) + geom_point())

(p <- ggplot(oDat, aes(geneExp, probeset)) + geom_point(position = position_jitter(height = 0.1)))
```
>Plot the expression level according to developmental stage.
```{r}
(p <- ggplot(oDat, aes(devStage, geneExp)) +  geom_point())
```
>Separate our 3 genes of interest with panels.
```{r}
(p <- p + facet_wrap(~ probeset))
```
>Add genotype information with color.
```{r}
(p <- p + aes(color = gType))
```
>Add in averages.
```{r}
(p <- p + stat_summary(fun.y = mean, geom = "point", shape = 4, size = 4))
```
Density plots
----------------------------------------------------------------------------
>Test out the 2 kinds of density plots in ggplot2: geom_density and stat_density.
```{r}
(p <- ggplot(oDat, aes(geneExp)) + geom_density())

(p <- ggplot(oDat, aes(geneExp)) + stat_density(geom = "line", position =
      "identity"))
```
>Adding our points in:
```{r}
(p <- ggplot(oDat, aes(geneExp)) + stat_density(geom = "line", position = "identity") + geom_point(aes(y = 0.05), position = position_jitter(height = 0.005)))
```
>Change the bandwidth.
```{r}
(p <- ggplot(oDat, aes(geneExp)) + stat_density(geom = "line", position = "identity", adjust = 0.5) + geom_point(aes(y = 0.05), position = position_jitter(height = 0.005)))
```
>Use panels to separate by genotype, or use color.
```{r}
(p <- p + facet_wrap(~ gType))

(p <- ggplot(oDat, aes(geneExp, color = gType)) + stat_density(geom = "line", position = "identity") + geom_point(aes(y = 0.05), position = position_jitter(height = 0.005)))

#You try!  Explore gene expression via developmental stage.
(p <- ggplot(oDat, aes(geneExp, color = devStage)) + stat_density(geom = "line", position = "identity") + geom_point(aes(y = 0.05), position = position_jitter(height = 0.005)))
```
Boxplots
------------------------------------------------------------------------------
>Make a boxplot.
```{r}
(p <- ggplot(oDat, aes(devStage, geneExp)) + geom_boxplot())
```
>Separate the genotypes with panels.
```{r}
(p <- p + facet_wrap(~ gType))
```
>Make a violin plot.
```{r}
(p <- ggplot(oDat, aes(devStage, geneExp)) + geom_violin())
```
Overplotting and plot matrix
---------------------------------------------------------------------------
>Pick two random samples to plot against each other from the large dataset. Put them in a data frame.
```{r}
set.seed(2)
(yo <- sample(1:ncol(prDat), size = 2))

bDat <- data.frame(y = prDat[[yo[1]]], z = prDat[[yo[2]]])
str(bDat)
```
>Plot samples against each other (scatterplot).
```{r}
(p <- ggplot(bDat, aes(z, y)) + geom_point())
```
Reduce the transparency of the data points (similar to smoothscatter).
```{r}
(p <- ggplot(bDat, aes(z, y)) + geom_point(alpha = 0.1))
```
>Plot using the 2D density function.
```{r}
(p <- ggplot(bDat, aes(z, y)) + stat_density2d())
```
>Use colors instead of lines to display the density.
```{r}
(p <- ggplot(bDat, aes(z, y)) + stat_density2d(geom = "tile", contour = F, aes(fill = ..density..)) + scale_fill_gradient(low = "white", high = "blue"))
```
>ggplot2 version of hexbin:
```{r}
(p <- ggplot(bDat, aes(z, y)) + stat_binhex())
```
>Larger sample size to address overplotting using plotmatrix.
```{r}
set.seed(3)
(yo <- sample(1:ncol(prDat), size = 4))
pairDat <- subset(prDat, select = yo)
str(pairDat)

(p <- plotmatrix(pairDat) + stat_binhex())
```
Heatmaps
-------------------------------------------------------------------------------
>Make a beautiful, beautiful heatmap after transforming data to tall format.  
```{r}
set.seed(1)
yo <- sample(1:nrow(prDat), size = 50)
hDat <- prDat[yo, ]
colnames(hDat) <- with(prDes, paste(devStage, gType, sidChar, sep = "_"))

prDatTall <- data.frame(sample = rep(colnames(hDat), each = nrow(hDat)),
                        probe = rownames(hDat), expression = unlist(hDat))

ggplot(prDatTall, aes(x = probe, y = sample, fill = expression)) +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
  geom_tile() +
  scale_fill_gradient2(low = jBuPuFun(256)[1], mid = jBuPuFun(256)[256/2],
                       high = jBuPuFun(256)[256],
                       midpoint = (max(prDatTall$expression) +
                       min(prDatTall$expression)) / 2, name = "Expression")
```
Exercise:
-------------------------------------------------------------------------------------------------------------------------------------
>The full photoRec dataset has 39 samples and 29949 probesets. Choose 2 ... or 20 ... or 200 random probesets/genes and look for gene expression differences between the two genotypes, wild type versus knockout. Make use of the graphing techniques discussed this week such as scatter plots, data heatmaps, correlation heatmaps, etc. 

Take a subset of 77 random probes and open a can of plotting whoopass.
```{r}
set.seed(777)
(exDat <- sample(1:nrow(prDat), size=77))
subDat <- prDat[exDat,]
str(subDat)
```
>Make a heatmap with lattice.
```{r}
hsubDat <- as.matrix(t(subDat))
rownames(hsubDat) <- with(prDes, paste(devStage, gType, sidChar, sep="_"))

heatmap(hsubDat,col = jBuPuFun(256),scale=c("column"),margins=c(5,8))
```
>Make a heatmap with ggplot2.
```{r}
colnames(subDat) <- with(prDes, paste(devStage, gType, sidChar, sep = "_"))

prDatTall <- data.frame(sample = rep(colnames(subDat), each = nrow(subDat)),
                        probe = rownames(subDat), expression = unlist(subDat))

ggplot(prDatTall, aes(x = probe, y = sample, fill = expression)) +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
  geom_tile() +
  scale_fill_gradient2(low = jBuPuFun(256)[1], mid = jBuPuFun(256)[256/2],
                       high = jBuPuFun(256)[256],
                       midpoint = (max(prDatTall$expression) +
                       min(prDatTall$expression)) / 2, name = "Expression")
```
>Make a density plot with lattice.
```{r}
resDat <- data.frame(sample = rep(names(subDat), each = 77), gene = LETTERS[1:77], geneExp = unlist(subDat))

Bw3 <- 0.5
n3 <- 500
densityplot(~geneExp, resDat,auto.key=TRUE, bw=Bw3, n=n3,
            main=paste("bw =",Bw3, ", n =",n3))
```

>Make a density plot with ggplot2.
```{r}
(p <- ggplot(resDat, aes(geneExp)) + stat_density(geom = "line", position = "identity") + geom_point(aes(y = 0.05), position = position_jitter(height = 0.005)))
```
>Be crazy and use splom().
```{r}
splom(subDat, panel=panel.smoothScatter, raster=TRUE)
```

