Seminar 7: RNA-Seq - Differential Expression Analysis
============================================================
>By Erica Acton

```{r include=FALSE}
library(edgeR)
library(splines)
library(DESeq)
library(limma)
library(VennDiagram)
library(ggplot2)
```

Load data and design files.  Check content.
```{r}
dat <- read.table("/home/eacton/R/stat540-2014-ACTON-ERICA/bottomly_count_table.tsv", header=TRUE, row.names=1)
des <- read.table("/home/eacton/R/stat540-2014-ACTON-ERICA/bottomly_phenodata.tsv", header=TRUE, row.names=1)
str(dat)
show(des)
all(rownames(des) == colnames(dat))
```
edgeR
-------------------------------------------------------------------------------------
Create groups according to sample strains.
```{r}
with(des, table(strain))
group <- factor(c(rep("1",10), rep("2",11)))
group
```
Create DGEList object of count data (matrix) and samples (data frame).
```{r}
dge.glm <- DGEList(counts = dat, group = group)
str(dge.glm)
names(dge.glm)
dge.glm[["samples"]]
nrow(dge.glm[[1]])
ncol(dge.glm[[1]])
```
Create design matrix.
```{r}
design <- model.matrix(~group)
design
```
Calculate the estimated common, trended, and tagwise dispersions.  Plot the tagwise dispersion agains log2-CPM.
```{r}
dge.glm.com.disp <- estimateGLMCommonDisp(dge.glm, design, verbose=TRUE)
dge.glm.trend.disp <- estimateGLMTrendedDisp(dge.glm.com.disp, design)
dge.glm.tag.disp <- estimateGLMTagwiseDisp(dge.glm.trend.disp, design)
plotBCV(dge.glm.tag.disp)
```
Fit a negative binomial generalized log-linear model to the read counts for each gene or transcript. Conduct genewise statistical tests.  Get top hits filtering for a FDR <0.01.
```{r}
fit <- glmFit(dge.glm.tag.disp,design)
colnames(coef(fit))
lrt <- glmLRT(fit,coef=2)
topTags(lrt)
tt.glm <- topTags(lrt, n=Inf)
class(tt.glm)
nrow(tt.glm$table[tt.glm$table$FDR<0.01,])
```
Grab interesting samples and their expression levels.
```{r}
interestingSamples <- rownames(tt.glm$table[tt.glm$table$FDR<1e-50,])
cpm(dge.glm.tag.disp)[interestingSamples,]
```
Get a summary of relative differential expression.  
```{r}
summary(de.glm <- decideTestsDGE(lrt, p=0.05, adjust="BH"))

#451 genes are under-expressed in Group 2 (Strain:DBA/2J) compared with Group 1 (Strain:CB/6J), 425 were over-expressed and 35660 showed no differences.
```
Plot the tagwise fold changes against log-CPM.
```{r}
tags.glm <- rownames(dge.glm.tag.disp)[as.logical(de.glm)]
plotSmear(lrt, de.tags=tags.glm)
abline(h=c(-2,2), col="blue")
```

Mini-Exercise (1): Filter to remove any gene that has counts equal to zero across all samples and repeat analysis.
-----------------------------------------------------------------------------------------
Remove any genes that has counts equal to zero across all samples.
```{r}
f1dat <- dat[rowSums(dat)!=0,]
str(f1dat)
```

Create groups according to sample strains.
```{r}
with(des, table(strain))
group <- factor(c(rep("1",10), rep("2",11)))
```
Create DGEList object of count data (matrix) and samples (data frame).
```{r}
f1dge.glm <- DGEList(counts = f1dat, group = group)
str(f1dge.glm)
names(f1dge.glm)
f1dge.glm[["samples"]]
nrow(f1dge.glm[[1]])
ncol(f1dge.glm[[1]])
```
Create design matrix.
```{r}
design <- model.matrix(~group)
design
```
Calculate the estimated common, trended, and tagwise dispersions.  Plot the tagwise dispersion agains log2-CPM.
```{r}
f1dge.glm.com.disp <- estimateGLMCommonDisp(f1dge.glm, design, verbose=TRUE)
f1dge.glm.trend.disp <- estimateGLMTrendedDisp(f1dge.glm.com.disp, design)
f1dge.glm.tag.disp <- estimateGLMTagwiseDisp(f1dge.glm.trend.disp, design)
plotBCV(f1dge.glm.tag.disp)
```
Fit a negative binomial generalized log-linear model to the read counts for each gene or transcript. Conduct genewise statistical tests.  Get top hits filtering for a FDR <0.01.
```{r}
f1fit <- glmFit(f1dge.glm.tag.disp,design)
colnames(coef(f1fit))
f1lrt <- glmLRT(f1fit,coef=2)
topTags(f1lrt)
f1tt.glm <- topTags(f1lrt, n=Inf)
class(f1tt.glm)
nrow(f1tt.glm$table[f1tt.glm$table$FDR<0.01,])
```
Grab interesting samples and their expression levels.
```{r}
f1interestingSamples <- rownames(f1tt.glm$table[f1tt.glm$table$FDR<1e-50,])
cpm(f1dge.glm.tag.disp)[f1interestingSamples,]
```
Get a summary of relative differential expression.  
```{r}
summary(f1de.glm <- decideTestsDGE(f1lrt, p=0.05, adjust="BH"))

#633 genes are under-expressed in Group 2 (Strain:DBA/2J) compared with Group 1 (Strain:CB/6J), 619 were over-expressed and 12680 showed no differences.
```
Plot the tagwise fold changes against log-CPM.
```{r}
f1tags.glm <- rownames(f1dge.glm.tag.disp)[as.logical(f1de.glm)]
plotSmear(f1lrt, de.tags=f1tags.glm)
abline(h=c(-2,2), col="blue")
```
DESeq
----------------------------------------------------------------------------------------
Get count and group information.
```{r}
deSeqDat <- newCountDataSet(dat, group)
head(counts(deSeqDat))
```
Estimate size factors to account for differences in library coverage, and estimate the variance.
```{r}
deSeqDat <- estimateSizeFactors(deSeqDat)
sizeFactors(deSeqDat)
```
Estimate and plot dispersions against the mean normalized counts.
```{r}

deSeqDat <- estimateDispersions(deSeqDat)
plotDispEsts(deSeqDat)
```
Fit the model and plot the log fold change against the mean normalized counts.
```{r}
deSeqresults <- nbinomTest(deSeqDat,levels(group)[1], levels(group)[2])
str(deSeqresults)
plotMA(deSeqresults)
```

Voom and Limma
-----------------------------------------------------------------------------------------
Calculate normalization factors to scale the raw library sizes.
```{r}
norm.factor <- calcNormFactors(dat)
```
Transform count data to log2-CPM and estimate and plot the mean-variance.  Compute weights. 

```{r}
dat.voomed <- voom(dat, design, plot=TRUE, lib.size=colSums(dat)*norm.factor)
dat.voomed
```
Compute linear model.
```{r}
fit <- lmFit(dat.voomed, design)
fit <- eBayes(fit)
topTable(fit)
```
Take Home Problem:
---------------------------------------------------------------------------------
Get genes differentially expressed at a threshold of an adjusted p value <1e-5 from voom. The number of genes obtained from voom analysis:
```{r}
voomtt <- topTable(fit, coef=2,n=Inf)
voomHits <-subset(voomtt, adj.P.Val<1e-5)
nrow(voomHits)
gene.ids <- rownames(voomHits)
voomHits <- cbind.data.frame(voomHits, gene.ids)

voomhits <- c(as.character(voomHits$gene.ids))
```
Get genes differentially expressed at a threshold of an adjusted p value <1e-5 from DESeq. The number of genes obtained from DESeq analysis:
```{r}
deSeqHits <- subset(deSeqresults, padj<1e-5)
nrow(deSeqHits)

deSeqresults <- c(as.character(deSeqHits$id))
```
Get genes differentially expressed at a threshold of an adjusted p value <1e-5 from edgeR. The number of genes obtained from edgeR analysis:
```{r}
edgetable <- tt.glm$table[tt.glm$table$FDR<1e-5,]
nrow(edgetable)

edgeRhits <- rownames(tt.glm$table[tt.glm$table$FDR<1e-5,])
```
A venn diagram of the overlap of hits for the 3 methods of analysis:
```{r}
allHits <- list(DESeq=deSeqresults, edgeR=edgeRhits, voom=voomhits)
plot.new()

venn.plot <- venn.diagram(allHits, filename=NULL, fill=c("red", "blue", "green"))
grid.draw(venn.plot)

```


